src/protocol/boolean/bit_decomposition.rs

```diff
@@ -0,0 +1,210 @@
+use super::bitwise_lt::BitwiseLessThan;
+use super::bitwise_sum::BitwiseSum;
+use crate::error::Error;
+use crate::ff::{Field, Int};
+use crate::protocol::boolean::local_secret_shared_bits;
+use crate::protocol::boolean::solved_bits::SolvedBits;
+use crate::protocol::context::{Context, SemiHonestContext};
+use crate::protocol::reveal::Reveal;
+use crate::protocol::RecordId;
+use crate::secret_sharing::Replicated;
+
+/// This is an implementation of "3. Bit-Decomposition" from I. Damgård et al..
+///
+/// It takes an input `[a] ∈ F_p` and outputs its bitwise additive share
+/// `[a]_B = ([a]_0,...,[a]_l-1)` where `[a]_i ∈ F_p`.
+///
+/// 3. Bit-Decomposition
+/// "Unconditionally Secure Constant-Rounds Multi-party Computation for Equality, Comparison, Bits, and Exponentiation"
+/// I. Damgård et al.
+pub struct BitDecomposition {}
+
+impl BitDecomposition {
+    #[allow(dead_code)]
+    pub async fn execute<F: Field>(
+        ctx: SemiHonestContext<'_, F>,
+        record_id: RecordId,
+        a_p: &Replicated<F>,
+    ) -> Result<Vec<Replicated<F>>, Error> {
+        // step 1 in the paper is just describing the input, `[a]_p` where `a ∈ F_p`
+
+        // Step 2. Generate random bitwise shares
+        let r = SolvedBits::execute(ctx.narrow(&Step::GenerateRandomBits), record_id).await?;
+        // Would like to do something like below. Where should I add the `RandomBitsGenerator`?
+        // let r = ctx.random_bits_generator.take_one().await?;
+        let r = r.expect("SolvedBits aborted");
+
+        // Step 3, 4. Reveal c = [a - b]_p
+        let c = ctx
+            .narrow(&Step::RevealAMinusB)
+            .reveal(record_id, &(a_p - &r.b_p))
+            .await?;
+        let c_b = local_secret_shared_bits(c.as_u128(), ctx.role());
+
+        // Step 5. Add back [b] bitwise. [d]_B = BitwiseSum(c, [b]_B) where d ∈ Z
+        //
+        // `BitwiseSum` outputs `l + 1` bits, so [d]_B is (l + 1)-bit long.
+        let d_b = BitwiseSum::execute(ctx.narrow(&Step::AddBtoC), record_id, &c_b, &r.b_b).await?;
+
+        // Step 6. p <? d
+        let p_b = local_secret_shared_bits(F::PRIME.into(), ctx.role());
+        let q_p = BitwiseLessThan::execute(ctx.narrow(&Step::IsPLessThanD), record_id, &p_b, &d_b)
+            .await?;
+
+        // Step 7. a bitwise scalar value `f_B = bits(2^l - p)`
+        let l = F::Integer::BITS;
+        let x = 2_u128.pow(l) - F::PRIME.into();
+        let f_b = (0..l).map(|i| F::from(x >> i & 1));
+
+        // Step 8, 9. [g_i] = [q] * f_i
+        let g_b = f_b
+            .into_iter()
+            .map(|f_bit| q_p.clone() * f_bit)
+            .collect::<Vec<_>>();
+
+        // Step 10. [h]_B = [d + g]_B, where [h]_B = ([h]_0,...[h]_(l+1))
+        //
+        // Again, `BitwiseSum` outputs `l + 1` bits. Since [d]_B is already
+        // `l + 1` bit long, [h]_B will be `l + 2`-bit long.
+        let h_b = BitwiseSum::execute(ctx.narrow(&Step::AddDtoG), record_id, &d_b, &g_b).await?;
```

Comment:
Yeah, I think we should be capable of saving ourselves some work here if we are only adding 5-bit values instead of 8-bit values (for the Fp31 case).

Reply:
OK, by shortening the bit lengths, we can reduce multiplications in BitwiseSum * 2, BitwiseLessThan. Is that correct?

If we start both [a] and [r] with 5-bit, that's 3 + 2 + 2 = 7 multiplications reduced per record for Fp31. As for Fp32BitPrime, we only save 1 multiplication?

Reply:
I have a suspicion that this is the cause of the bitwise shares sometimes being larger than the prime.